README.md
# serial_csv

[![pub package](https://img.shields.io/pub/v/serial_csv.svg)](https://pub.dev/packages/serial_csv)
![ci](https://github.com/Tienisto/serial_csv/actions/workflows/ci.yml/badge.svg)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

High performance CSV encoder and decoder with retained types and nullability.

## Introduction

This package is intended to quickly encode and decode csv while keeping all type information.

Compared to JSON, CSV is a much more compact format that is also readable without having to "beautify" it.

Instead of supporting the entire CSV specification, this package uses a more strict format that is optimized for performance.

The following data types are preserved: `String`, `int`, `double`, `bool`, `null`.

## Benchmark

The benchmark consists of 30 billion cells (5000 rows x 3000 columns x 2000 iterations) running on a Ryzen 5 5600X.

For comparison, the packages [csv](https://pub.dev/packages/csv) (v5.0.2),
[fast_csv](https://pub.dev/packages/fast_csv) (v0.1.44),
and [csvwriter](https://pub.dev/packages/csvwriter) (v1.3.0) were used.

![benchmark](https://raw.githubusercontent.com/Tienisto/serial_csv/main/assets/benchmark.png)

\* `jsonEncode` and `jsonDecode` are using an equivalent data representation.

You can run the benchmark yourself with this [code example](https://github.com/Tienisto/serial_csv/blob/main/example/benchmark.dart).

## CSV structure

This library uses a more strict CSV format to achieve high performance.

- string cells are always double-quoted
- numeric and boolean cells are not double-quoted
- null cells are encoded as empty cells
- double-quotes within cells are escaped with another double-quote
- cells are separated by commas
- every row must be terminated with `\n` (LF)
- the last row must be terminated with `\n` (LF)
- no comments (`#`) are allowed

There is no "vendor lock-in" as the CSV is still valid CSV. You can use any other CSV library to read the CSV.

Just be careful not to modify the CSV with other tools as you might break the strict format.

## Usage

Use `SerialCsv.encode` and `SerialCsv.decode` to encode and decode CSV.

```csv
"a","b","c"
1,2.3,"3"
4,,true
```

```dart
String encoded = SerialCsv.encode([
  ['a', 'b', 'c'],
  [1, 2.3, '3'],
  ['4', null, true],
]);

List<List<dynamic>> decoded = SerialCsv.decode(encoded);

// The type information is retained.
String s = decoded[0][0]; // 'a'
int i = decoded[1][0]; // 1
double d = decoded[1][1]; // 2.3
bool b = decoded[2][2]; // true
Object? n = decoded[2][1]; // null
```

## Strings only

Encode and decode a list of rows with only string cells. No `null` is allowed.

```dart
const rows = [
  ['a', 'b', 'c'],
  ['1', '2', '3'],
  ['4', '5', 'true'],
];

String encoded = SerialCsv.encodeStrings(rows);
List<List<String>> decoded = SerialCsv.decodeStrings(encoded);
```

## Key-Value

Encode and decode a non-nested map. Keys must be strings. Values can be any supported type.

```dart
const map = {
  'a': '1',
  'b': 2,
  'c': 3.4,
  'd': true,
  'e': null,
};

String encoded = SerialCsv.encodeMap(map);
Map<String, dynamic> decoded = SerialCsv.decodeMap(encoded);

// The type information is retained.
String s = decoded['a']; // '1'
int i = decoded['b']; // 2
double d = decoded['c']; // 3.4
bool b = decoded['d']; // true
Object? n = decoded['e']; // null
```


CHANGELOG.md
## 0.4.0

- add specialized `encode` and `decode` methods for `int`, `double` and `bool`
- rename `encodeStringList` and `decodeStringList` to `encodeStrings` and `decodeStrings`
- significantly improve parsing performance

## 0.3.0

- make dart constraints less strict (requires only dart 2.12+)
- add `encodeMap` and `decodeMap` methods

## 0.2.1

- fix benchmark image url

## 0.2.0

- method signature changed from `List<List<Object?>>` to `List<List<dynamic>>`
- add `encodeStringList` and `decodeStringList` methods

## 0.1.0

- initial release


test/csv_decoder_test.dart
import 'package:serial_csv/serial_csv.dart';
import 'package:test/test.dart';

void main() {
  group('SerialCsvDecoder.decode', () {
    test('Should decode single string', () {
      final result = SerialCsv.decode('"a"\n');

      expect(result, [
        ['a'],
      ]);
    });

    test('Should decode empty string', () {
      final result = SerialCsv.decode('""\n');

      expect(result, [
        ['']
      ]);
    });

    test('Should decode multiple empty strings', () {
      final result = SerialCsv.decode('"","",""\n""\n""\n');

      expect(result, [
        ['', '', ''],
        [''],
        [''],
      ]);
    });

    test('Should decode with nulls', () {
      final result = SerialCsv.decode('''"a",,"null"
1,"null",
,,
''');

      expect(result, [
        ['a', null, 'null'],
        [1, 'null', null],
        [null, null, null],
      ]);
    });

    test('Should decode with quotes', () {
      final result = SerialCsv.decode('"a","b","""c""a"\n');

      expect(result, [
        ['a', 'b', '"c"a'],
      ]);
    });

    test('Should decode with commas', () {
      final result = SerialCsv.decode('"a","b","c,d"\n');

      expect(result, [
        ['a', 'b', 'c,d'],
      ]);
    });

    test('Should decode with newlines', () {
      final result = SerialCsv.decode('"a","b","c\nd"\n');

      expect(result, [
        ['a', 'b', 'c\nd'],
      ]);
    });

    test('Should decode surrogate pairs', () {
      final result = SerialCsv.decode('"a","b","c𠀀d"\n');

      expect(result, [
        ['a', 'b', 'c𠀀d'],
      ]);
    });

    test('Should decode (integration test)', () {
      final result = SerialCsv.decode('''"a",,"c"
1,2.3,true
,,,
-3.22
"hello""world"
false,4,""
-2.3e-4
''');

      expect(result, [
        ['a', null, 'c'],
        [1, 2.3, true],
        [null, null, null, null],
        [-3.22],
        ['hello"world'],
        [false, 4, ''],
        [-2.3e-4],
      ]);
    });
  });

  group('SerialCsvDecoder.decodeStrings', () {
    test('Should decode single string', () {
      final result = SerialCsv.decodeStrings('"a"\n');

      expect(result, [
        ['a'],
      ]);
    });

    test('Should decode with quotes', () {
      final result = SerialCsv.decodeStrings('"a","b","""c""a"\n');

      expect(result, [
        ['a', 'b', '"c"a'],
      ]);
    });

    test('Should decode with commas', () {
      final result = SerialCsv.decodeStrings('"a","b","c,d"\n');

      expect(result, [
        ['a', 'b', 'c,d'],
      ]);
    });

    test('Should decode with newlines', () {
      final result = SerialCsv.decodeStrings('"a","b","c\nd"\n');

      expect(result, [
        ['a', 'b', 'c\nd'],
      ]);
    });

    test('Should decode surrogate pairs', () {
      final result = SerialCsv.decodeStrings('"a","b","c𠀀d"\n');

      expect(result, [
        ['a', 'b', 'c𠀀d'],
      ]);
    });

    test('Should decode (integration test)', () {
      final result = SerialCsv.decodeStrings('''"a","b11","c"
"ddd","ee","fff"
"g","h","i"
"hello""world"
''');

      expect(result, [
        ['a', 'b11', 'c'],
        ['ddd', 'ee', 'fff'],
        ['g', 'h', 'i'],
        ['hello"world'],
      ]);
    });
  });

  group('SerialCsvDecoder.decodeIntegers', () {
    test('Integration test', () {
      final result = SerialCsv.decodeIntegers('''123124,34634,234249
3326,-345,3
235,358,-44
987
''');

      expect(result, [
        [123124, 34634, 234249],
        [3326, -345, 3],
        [235, 358, -44],
        [987],
      ]);
    });
  });

  group('SerialCsvDecoder.decodeDoubles', () {
    test('Integration test', () {
      final result = SerialCsv.decodeDoubles('''123124.234,34634.234,234249.234
3326.234,-345.234,3.234
235.234,358.234,-44.234
987.234
''');

      expect(result, [
        [123124.234, 34634.234, 234249.234],
        [3326.234, -345.234, 3.234],
        [235.234, 358.234, -44.234],
        [987.234],
      ]);
    });
  });

  group('SerialCsvDecoder.decodeBooleans', () {
    test('Integration test', () {
      final result = SerialCsv.decodeBooleans('''true,false,true
false,true,false
true,true,false
false
''');

      expect(result, [
        [true, false, true],
        [false, true, false],
        [true, true, false],
        [false],
      ]);
    });
  });

  group('SerialCsvDecoder.decodeMap', () {
    test('Should decode a simple map', () {
      final result = SerialCsv.decodeMap('"a","aa"\n');

      expect(result, {
        'a': 'aa',
      });
    });

    test('Should decode normally', () {
      final result = SerialCsv.decodeMap('''"a","aa"
"b",2345
"c",23.45
"d",true
"e",false
"f",
''');

      expect(result, {
        'a': 'aa',
        'b': 2345,
        'c': 23.45,
        'd': true,
        'e': false,
        'f': null,
      });
      expect(result['b'], isA<int>());
      expect(result['c'], isA<double>());
    });

    test('Should decode mupltiple entries', () {
      final result = SerialCsv.decodeMap('"a","11"\n"b","22"\n');

      expect(result, {
        'a': '11',
        'b': '22',
      });
    });

    test('Should decode with quotes', () {
      final result = SerialCsv.decodeMap('"a","""b"""\n"""c""","d"\n');

      expect(result, {
        'a': '"b"',
        '"c"': 'd',
      });
    });

    test('Should decode with commas', () {
      final result = SerialCsv.decodeMap('"a","b,c"\n');

      expect(result, {
        'a': 'b,c',
      });
    });

    test('Should decode with newlines', () {
      final result = SerialCsv.decodeMap('"a","b\nc"\n');

      expect(result, {
        'a': 'b\nc',
      });
    });

    test('Should decode surrogate pairs', () {
      final result = SerialCsv.decodeMap('"a","b𠀀c"\n');

      expect(result, {
        'a': 'b𠀀c',
      });
    });
  });
}


test/csv_encoder_test.dart
import 'package:serial_csv/serial_csv.dart';
import 'package:test/test.dart';

void main() {
  group('SerialCsv.encode', () {
    test('Should encode normally', () {
      final result = SerialCsv.encode([
        ['a', 'b', 'c'],
        ['1', '2', '3'],
      ]);

      expect(result, '"a","b","c"\n"1","2","3"\n');
    });

    test('Should encode with quotes', () {
      final result = SerialCsv.encode([
        ['a', 'b', '"c"'],
      ]);

      expect(result, '"a","b","""c"""\n');
    });

    test('Should encode with newlines', () {
      final result = SerialCsv.encode([
        ['a', 'b', 'c\nd'],
      ]);

      expect(result, '"a","b","c\nd"\n');
    });

    test('Should not add double-quotes for numbers and booleans', () {
      final result = SerialCsv.encode([
        ['a', 'b', 'c'],
        [1, 2.3, true],
      ]);

      expect(result, '"a","b","c"\n1,2.3,true\n');
    });

    test('Should encode with nulls', () {
      final result = SerialCsv.encode([
        ['a', 'b', null],
      ]);

      expect(result, '"a","b",\n');
    });
  });

  group('SerialCsv.encodeStrings', () {
    test('Should encode normally', () {
      final result = SerialCsv.encodeStrings([
        ['a', 'b', 'c'],
        ['1', '2', '3'],
      ]);

      expect(result, '"a","b","c"\n"1","2","3"\n');
    });

    test('Should encode with quotes', () {
      final result = SerialCsv.encodeStrings([
        ['a', 'b', '"c"'],
      ]);

      expect(result, '"a","b","""c"""\n');
    });

    test('Should encode with newlines', () {
      final result = SerialCsv.encodeStrings([
        ['a', 'b', 'c\nd'],
      ]);

      expect(result, '"a","b","c\nd"\n');
    });
  });

  group('SerialCsv.encodeMap', () {
    test('Should encode normally', () {
      final result = SerialCsv.encodeMap({
        'a': '1',
        'b': 2,
        'c': true,
        'd': null,
        'e': 'hello"world',
      });

      expect(result, '"a","1"\n"b",2\n"c",true\n"d",\n"e","hello""world"\n');
    });
  });

  group('SerialCsv.encodeGeneric', () {
    test('Integer integration test', () {
      final result = SerialCsv.encodeIntegers([
        [1, 2, 3],
        [],
        [-4, 0],
        [1000]
      ]);

      expect(result, '1,2,3\n\n-4,0\n1000\n');
    });

    test('Double integration test', () {
      final result = SerialCsv.encodeDoubles([
        [1.0, 2.0, 3.0],
        [],
        [-4.0, 0.0],
        [1000.0]
      ]);

      expect(result, '1.0,2.0,3.0\n\n-4.0,0.0\n1000.0\n');
    });

    test('Boolean integration test', () {
      final result = SerialCsv.encodeBooleans([
        [true, false, true],
        [],
        [false, true],
        [true]
      ]);

      expect(result, 'true,false,true\n\nfalse,true\ntrue\n');
    });
  });
}


lib/serial_csv.dart
import 'package:serial_csv/src/csv_decoder.dart';
import 'package:serial_csv/src/csv_encoder.dart';

class SerialCsv {
  SerialCsv._();

  /// Encodes a list of rows into a csv string.
  static String encode(List<List<dynamic>> data) {
    return const SerialCsvEncoder().encode(data);
  }

  /// A specialized version of encode that encodes a list of rows of strings.
  static String encodeStrings(List<List<String>> data) {
    return const SerialCsvEncoder().encodeStrings(data);
  }

  /// A specialized version of encode that encodes a list of rows of integers.
  static String encodeIntegers(List<List<int>> data) {
    return const SerialCsvEncoder().encodeGeneric<int>(data);
  }

  /// A specialized version of encode that encodes a list of rows of doubles.
  static String encodeDoubles(List<List<double>> data) {
    return const SerialCsvEncoder().encodeGeneric<double>(data);
  }

  /// A specialized version of encode that encodes a list of rows of booleans.
  static String encodeBooleans(List<List<bool>> data) {
    return const SerialCsvEncoder().encodeGeneric<bool>(data);
  }

  /// Encodes a map of key-value pairs into a csv string.
  /// The encoded csv will have two columns.
  static String encodeMap(Map<String, dynamic> map) {
    return const SerialCsvEncoder().encodeMap(map);
  }

  /// Decodes a csv string into a list of rows.
  static List<List<dynamic>> decode(String csv) {
    return const SerialCsvDecoder().decode(csv);
  }

  /// A specialized version of decode that returns a list of rows of strings.
  static List<List<String>> decodeStrings(String csv) {
    return const SerialCsvDecoder().decode(csv).cast();
  }

  /// A specialized version of decode that returns a list of rows of integers.
  static List<List<int>> decodeIntegers(String csv) {
    return const SerialCsvDecoder().decodeIntegers(csv);
  }

  /// A specialized version of decode that returns a list of rows of doubles.
  static List<List<double>> decodeDoubles(String csv) {
    return const SerialCsvDecoder().decodeDoubles(csv);
  }

  /// A specialized version of decode that returns a list of rows of booleans.
  static List<List<bool>> decodeBooleans(String csv) {
    return const SerialCsvDecoder().decodeBooleans(csv);
  }

  /// Decodes a CSV consisting of key-value pairs into a map.
  /// The csv must have two columns and the first column must be a string.
  static Map<String, dynamic> decodeMap(String csv) {
    return const SerialCsvDecoder().decodeMap(csv);
  }
}


lib/src/csv_encoder.dart
final _buffer = StringBuffer();

class SerialCsvEncoder {
  const SerialCsvEncoder();

  /// Encodes a list of rows into a csv string.
  String encode(List<List<dynamic>> data) {
    for (final row in data) {
      for (int i = 0; i < row.length; i++) {
        _buffer.write(_encodeCell(row[i]));

        if (i != row.length - 1) {
          _buffer.write(',');
        }
      }
      _buffer.write('\n');
    }
    final result = _buffer.toString();
    _buffer.clear();
    return result;
  }

  String _encodeCell(Object? cell) {
    if (cell == null) {
      return '';
    } else if (cell is String) {
      return '"${cell.replaceAll('"', '""')}"';
    } else {
      return cell.toString();
    }
  }

  /// Encodes a list of rows into a csv string.
  String encodeStrings(List<List<String>> data) {
    for (final row in data) {
      for (int i = 0; i < row.length; i++) {
        _buffer.write('"');
        _buffer.write(row[i].replaceAll('"', '""'));

        if (i != row.length - 1) {
          _buffer.write('",');
        } else {
          _buffer.write('"');
        }
      }
      _buffer.write('\n');
    }
    final result = _buffer.toString();
    _buffer.clear();
    return result;
  }

  /// Encodes a list of rows into a csv string.
  /// Suitable for integers, doubles, and booleans.
  String encodeGeneric<T>(List<List<T>> data) {
    for (final row in data) {
      for (int i = 0; i < row.length; i++) {
        _buffer.write(row[i].toString());

        if (i != row.length - 1) {
          _buffer.write(',');
        }
      }
      _buffer.write('\n');
    }
    final result = _buffer.toString();
    _buffer.clear();
    return result;
  }

  /// Encodes a map into a csv string.
  String encodeMap(Map<String, dynamic> map) {
    for (final key in map.keys) {
      _buffer.write('"');
      _buffer.write(key.replaceAll('"', '""'));
      _buffer.write('",');
      _buffer.write(_encodeCell(map[key]));
      _buffer.write('\n');
    }
    final result = _buffer.toString();
    _buffer.clear();
    return result;
  }
}


example/benchmark.dart
import 'dart:convert';
import 'dart:math';

import 'package:csv/csv.dart';
import 'package:csvwriter/csvwriter.dart';
import 'package:serial_csv/serial_csv.dart';
import 'package:fast_csv/fast_csv.dart' as fast_csv;

void main() {
  benchmarkParseStrings();
}

List<List> getTypedList() {
  return List.generate(
    1000,
    (index) => [
      ['ae5szjmz"nteje3wT', 'bwZshH!?dhdrj', '${'ka$index'.hashCode.toRadixString(16)}!'],
      [index.hashCode, 2.43, '-33'],
      [-4, null, true],
      [null, null, null],
      [false, index % 2 == 0, false],
    ],
  ).expand((element) => element).toList();
}

List<List<String>> getStringList() {
  return List.generate(
    1000,
    (index) => [
      ['ae5szjmz"nteje3wT', 'bwZshH!?dhdrj', '${'ka$index'.hashCode.toRadixString(16)}!'],
      ['aefew4t9z438"goGHUIHUIrqwe', 'RJ%U§)(u35wfesqqq', 'DEOFAFSAVka2411$index bb'],
      ['oas8fc9FU()AÖPPfa3JJKL§"" ', 'öwpod9elKKS${'gg$index'.hashCode}', '12312413432'],
      ['rODEO§)FLlFAQ!!!!&&&&', 'LLFWOPOSAPKOPWFF', 'aldsi9OP§FLSDNFKLew3332'],
      ['fsdfeoi${'aaa$index'.hashCode}---', 'owejftiw04380', 'pfojw0t9j03ow4pzjg9psw548gjw54'],
    ],
  ).expand((element) => element).toList();
}

List<List<int>> getIntList() {
  final random = Random();
  return List.generate(
    1000,
    (index) => List.generate(
      3,
      (index) => random.nextInt(1000) - 500,
    ),
  );
}

List<List<double>> getDoubleList() {
  final random = Random();
  return List.generate(
    1000,
    (index) => List.generate(
      3,
      (index) => random.nextDouble() * 1000 - 500.0,
    ),
  );
}

Map<String, dynamic> getMap() {
  return Map<String, dynamic>.fromEntries(List.generate(2000, (index) {
    final String key = 'k$index.hashCode'.hashCode.toRadixString(16);
    final dynamic value;
    switch (index % 6) {
      case 0:
        value = index.hashCode;
        break;
      case 1:
        value = index.hashCode.toRadixString(16);
        break;
      case 2:
        value = index.hashCode / 2;
        break;
      case 3:
        value = index.hashCode % 2 == 0;
        break;
      default:
        value = null;
        break;
    }
    return MapEntry(key, value);
  }));
}

void benchmarkParseTyped() {
  final inputStructured = getTypedList();
  final inputCsv = SerialCsv.encode(inputStructured);
  const iterations = 2000;

  _heatup(
    name: 'parse mixed types',
    iterations: iterations,
    functions: [
      () => SerialCsv.decode(inputCsv),
      () => fast_csv.parse(inputCsv),
      () => const CsvToListConverter(eol: '\n').convert(inputCsv),
      () => jsonDecode(jsonEncode(inputStructured)),
    ],
  );

  _benchmark(
    name: 'SerialCsv.decode',
    iterations: iterations,
    func: () => SerialCsv.decode(inputCsv),
  );

  _benchmark(
    name: 'fast_csv',
    iterations: iterations,
    func: () => fast_csv.parse(inputCsv),
  );

  _benchmark(
    name: 'csv',
    iterations: iterations,
    func: () => const CsvToListConverter(eol: '\n').convert(inputCsv),
  );

  final encoded = jsonEncode(inputStructured); // encode only once
  _benchmark(
    name: 'jsonDecode',
    iterations: iterations,
    func: () => jsonDecode(encoded),
  );
}

void benchmarkParseStrings() {
  final inputStructured = getStringList();
  final inputCsv = SerialCsv.encode(inputStructured);
  const iterations = 2000;

  _heatup(
    name: 'parse strings',
    iterations: iterations,
    functions: [
      () => SerialCsv.decode(inputCsv),
      () => SerialCsv.decodeStrings(inputCsv),
      () => fast_csv.parse(inputCsv),
      () => const CsvToListConverter(eol: '\n').convert(inputCsv),
      () => jsonDecode(jsonEncode(inputStructured)),
    ],
  );

  _benchmark(
    name: 'SerialCsv.decode',
    iterations: iterations,
    func: () => SerialCsv.decode(inputCsv),
  );

  _benchmark(
    name: 'SerialCsv.decodeStrings',
    iterations: iterations,
    func: () => SerialCsv.decodeStrings(inputCsv),
  );

  _benchmark(
    name: 'fast_csv',
    iterations: iterations,
    func: () => fast_csv.parse(inputCsv),
  );

  _benchmark(
    name: 'csv',
    iterations: iterations,
    func: () => const CsvToListConverter(eol: '\n').convert(inputCsv),
  );

  final encoded = jsonEncode(inputStructured); // encode only once
  _benchmark(
    name: 'jsonDecode',
    iterations: iterations,
    func: () => jsonDecode(encoded),
  );
}

void benchmarkParseInt() {
  final inputStructured = getIntList();
  final inputCsv = SerialCsv.encode(inputStructured);
  const iterations = 2000;

  _heatup(
    name: 'parse ints',
    iterations: iterations,
    functions: [
      () => SerialCsv.decode(inputCsv),
      () => SerialCsv.decodeIntegers(inputCsv),
      () => fast_csv.parse(inputCsv),
      () => const CsvToListConverter(eol: '\n').convert(inputCsv),
      () => jsonDecode(jsonEncode(inputStructured)),
    ],
  );

  _benchmark(
    name: 'SerialCsv.decode',
    iterations: iterations,
    func: () => SerialCsv.decode(inputCsv).cast<List<int>>(),
  );

  _benchmark(
    name: 'SerialCsv.decodeIntegers',
    iterations: iterations,
    func: () => SerialCsv.decodeIntegers(inputCsv),
  );

  _benchmark(
    name: 'fast_csv',
    iterations: iterations,
    func: () => fast_csv.parse(inputCsv),
  );

  _benchmark(
    name: 'csv',
    iterations: iterations,
    func: () => const CsvToListConverter(eol: '\n').convert(inputCsv),
  );

  final encoded = jsonEncode(inputStructured); // encode only once
  _benchmark(
    name: 'jsonDecode',
    iterations: iterations,
    func: () => jsonDecode(encoded),
  );
}

void benchmarkParseDouble() {
  final inputStructured = getDoubleList();
  final inputCsv = SerialCsv.encode(inputStructured);
  const iterations = 2000;

  _heatup(
    name: 'parse doubles',
    iterations: iterations,
    functions: [
      () => SerialCsv.decode(inputCsv),
      () => SerialCsv.decodeDoubles(inputCsv),
      () => jsonDecode(jsonEncode(inputStructured)),
    ],
  );

  _benchmark(
    name: 'SerialCsv.decode',
    iterations: iterations,
    func: () => SerialCsv.decode(inputCsv).cast<List<double>>(),
  );

  _benchmark(
    name: 'SerialCsv.decodeDoubles',
    iterations: iterations,
    func: () => SerialCsv.decodeDoubles(inputCsv),
  );

  final encoded = jsonEncode(inputStructured); // encode only once
  _benchmark(
    name: 'jsonDecode',
    iterations: iterations,
    func: () => jsonDecode(encoded),
  );
}

void benchmarkParseMap() {
  final map = getMap();
  final inputCsv = SerialCsv.encodeMap(map);
  const iterations = 2000;

  _heatup(
    name: 'parseMap',
    iterations: iterations,
    functions: [
      () => SerialCsv.decodeMap(inputCsv),
      () => fast_csv.parse(inputCsv),
      () => jsonDecode(jsonEncode(map)),
    ],
  );

  _benchmark(
    name: 'SerialCsv.decodeMap',
    iterations: iterations,
    func: () => SerialCsv.decodeMap(inputCsv),
  );

  _benchmark(
    name: 'SerialCsv.decode + map conversion',
    iterations: iterations,
    func: () {
      final result = SerialCsv.decode(inputCsv);
      return {
        for (var i = 0; i < result.length; i++) result[i][0]: result[i][1],
      };
    },
  );

  _benchmark(
    name: 'fast_csv (rows representation)',
    iterations: iterations,
    func: () => fast_csv.parse(inputCsv),
  );

  _benchmark(
    name: 'csv (rows representation)',
    iterations: iterations,
    func: () => const CsvToListConverter(eol: '\n').convert(inputCsv),
  );

  final encodedJson = jsonEncode(map);
  _benchmark(
    name: 'jsonDecode',
    iterations: iterations,
    func: () => jsonDecode(encodedJson),
  );
}

void benchmarkEncode() {
  final input = getTypedList();
  const iterations = 2000;

  _heatup(
    name: 'encode mixed types',
    iterations: iterations,
    functions: [
      () => SerialCsv.encode(input),
      () => const ListToCsvConverter().convert(input),
      () => jsonEncode(input),
    ],
  );

  _benchmark(
    name: 'SerialCsv.encode',
    iterations: iterations,
    func: () => SerialCsv.encode(input),
  );

  _benchmark(
    name: 'csvwriter',
    iterations: iterations,
    func: () {
      final buffer = StringBuffer();
      final writer = CsvWriter(buffer, 5, endOfLine: '\n');
      for (final row in input) {
        writer.writeData(data: row);
      }
      return buffer.toString();
    },
  );

  _benchmark(
    name: 'csv',
    iterations: iterations,
    func: () => const ListToCsvConverter().convert(input),
  );

  _benchmark(
    name: 'jsonEncode',
    iterations: iterations,
    func: () => jsonEncode(input),
  );
}

void benchmarkEncodeStrings() {
  final input = getStringList();
  const iterations = 2000;

  _heatup(
    name: 'encode strings',
    iterations: iterations,
    functions: [
      () => SerialCsv.encodeStrings(input),
      () => const ListToCsvConverter().convert(input),
    ],
  );

  _benchmark(
    name: 'SerialCsv.encodeStrings',
    iterations: iterations,
    func: () => SerialCsv.encodeStrings(input),
  );

  _benchmark(
    name: 'csvwriter',
    iterations: iterations,
    func: () {
      final buffer = StringBuffer();
      final writer = CsvWriter(buffer, 5, endOfLine: '\n');
      for (final row in input) {
        writer.writeData(data: row);
      }
      return buffer.toString();
    },
  );

  _benchmark(
    name: 'jsonEncode',
    iterations: iterations,
    func: () => jsonEncode(input),
  );

  _benchmark(
    name: 'csv',
    iterations: iterations,
    func: () => const ListToCsvConverter().convert(input),
  );
}

void benchmarkEncodeMap() {
  final map = getMap();
  const iterations = 3000;

  _heatup(
    name: 'encode map',
    iterations: iterations,
    functions: [
      () => SerialCsv.encodeMap(map),
      () => jsonEncode(map),
    ],
  );

  _benchmark(
    name: 'SerialCsv.encodeMap',
    iterations: iterations,
    func: () => SerialCsv.encodeMap(map),
  );

  _benchmark(
    name: 'jsonEncode',
    iterations: iterations,
    func: () => jsonEncode(map),
  );
}

void _benchmark<T>({
  required String name,
  required int iterations,
  required T Function() func,
  bool printResult = false,
}) {
  if (printResult) {
    final result = func();
    if (result is List<List>) {
      _printTypes2(result);
    } else if (result is List) {
      _printTypes(result);
    } else {
      print('Print result:');
      print(result.runtimeType);
      print(result);
    }
    return;
  }

  final stopwatch = Stopwatch()..start();
  for (int i = 0; i < iterations; i++) {
    func();
  }
  stopwatch.stop();
  print('[${stopwatch.elapsedMilliseconds.toString().padLeft(5)} ms] $name');
}

void _heatup<T>({
  required String name,
  required int iterations,
  required List<T Function()> functions,
  bool printResult = false,
}) {
  print('Starting heat-up for "$name" ...');

  if (printResult) {
    print(functions.first());
    return;
  }

  final stopwatch = Stopwatch()..start();
  for (final fun in functions) {
    for (int i = 0; i < iterations; i++) {
      fun();
    }
  }
  stopwatch.stop();
  print('[${stopwatch.elapsedMilliseconds.toString().padLeft(5)} ms] heat-up');
}

void _printTypes(List<dynamic> result) {
  for (int i = 0; i < result.length; i++) {
    final row = result[i];
    if (row is List) {
      for (int j = 0; j < row.length; j++) {
        print('decoded[$i][$j]: ${row[j]} (${row[j].runtimeType})');
      }
    } else {
      print('decoded[$i]: $row (${row.runtimeType})');
    }
  }
}

void _printTypes2(List<List<dynamic>> result) {
  for (int row = 0; row < result.length; row++) {
    for (int col = 0; col < result[row].length; col++) {
      print('decoded[$row][$col]: ${result[row][col]} (${result[row][col].runtimeType})');
    }
  }
}
